neuraxle/distributed/streaming.py

```diff
@@ -0,0 +1,511 @@
+# (step_name, step)
+QueuedPipelineStepsTuple = Union[
+    Tuple[str, int, BaseStep],
+    Tuple[str, int, int, BaseStep],
+    Tuple[str, int, List[Tuple], BaseStep],
+    Tuple[str, int, List[Tuple], BaseStep],
+    Tuple[str, BaseStep]
+]
+
+
+class BaseQueuedPipeline(NonFittableMixin, CustomPipelineMixin, Pipeline):
+    """
+    Sub class of :class:`Pipeline`.
+    Transform data in many pipeline steps at once in parallel in the pipeline using multiprocessing Queues.
+
+    Example usage :
+
+    .. code-block:: python
+
+        # step name, step
+
+        p = QueuedPipeline([
+            ('step_a', Identity()),
+            ('step_b', Identity()),
+        ], n_workers=1, batch_size=10, max_size=10)
+
+        # number of workers
+
+        p = QueuedPipeline([
+            ('step_a', 1, Identity()),
+            ('step_b', 1, Identity()),
+        ], batch_size=10, max_size=10)
+
+        # number of workers, and max size
+
+        p = QueuedPipeline([
+            ('step_a', 1, 10, Identity()),
+            ('step_b', 1, 10, Identity()),
+        ], batch_size=10)
+
+        # number of workers for each step, and additional argument for each worker
+
+        p = QueuedPipeline([
+            ('step_a', 1, [('host', 'host1'), ('host', 'host2')], 10, Identity())
+        ], batch_size=10)
+
+        # number of workers for each step, additional argument for each worker, and max size
+
+        p = QueuedPipeline([
+            ('step_a', 1, [('host', 'host1'), ('host', 'host2')], 10, Identity())
+        ], batch_size=10)
+
+
+    .. seealso::
+        :class:`QueueWorker`,
+        :class:`QueueJoiner`,
+        :class:`CustomPipelineMixin`,
+        :class:`Pipeline`
+    """
+
+    def __init__(
+            self,
+            steps: List[QueuedPipelineStepsTuple],
+            batch_size,
+            n_workers_per_step=None,
+            max_size=None,
+            data_joiner=None,
+            use_threading=False,
+            cache_folder=None
+    ):
+        NonFittableMixin.__init__(self)
+        CustomPipelineMixin.__init__(self)
+
+        if data_joiner is None:
+            data_joiner = NumpyConcatenateOuterBatch()
+        self.data_joiner = data_joiner
+        self.max_size = max_size
+        self.batch_size = batch_size
+        self.n_workers_per_step = n_workers_per_step
+        self.use_threading = use_threading
+
+        Pipeline.__init__(self, steps=self._initialize_steps_as_tuple(steps), cache_folder=cache_folder)
+
+    def _initialize_steps_as_tuple(self, steps):
+        """
+        Wrap each step by a :class:`QueueWorker` to  allow data to flow in many pipeline steps at once in parallel.
+
+        :param steps: (name, n_workers, step)
+        :type steps: NameNWorkerStepTupleList
+        :return: steps as tuple
+        :rtype: NamedTupleList
+        """
+        steps_as_tuple: NamedTupleList = []
+        for step in steps:
+            queue_worker = self._create_queue_worker(step)
```

Comment:
we need a custom saver that removes multiprocess queue...

Reply:
mount(), unmount()

Reply:
Let's open an issue for all what regards saving as a 2nd pr to do after in `0.4.1`.

Reply:
https://github.com/Neuraxio/Neuraxle/issues/307